File: Chapter-04-Strings-AssociativeArrays/String-Split/String-Split.py
# Write your own function to recreate the string.split() function of python. python split works a little differently compared to javascript.

# Splits a string into an array of substrings. It takes in a seperator based on which to split the string. If the seperator is not specified, it splits the words present in the string (this behaviour is different than in javascript). It accepts a second parm which limits the number of splits (this way second parm is handled is different from js). It doesn't handle an empty string as seperator (different from js)

# 'stringstreet'.split() returns ['stringstreet']; 'string street'.split() returns ['string', 'street']; 'stringstreet'.split('t') returns ['s', 'rings', 'ree', '']; 'stringstreet'.split('t', 2) returns ['s', 'rings', 'reet']; 'stringstreet'.split('t', 1) returns ['s', 'ringstreet']



def split(str, sep = False, limit = False):
    # handle empty seperator
    if (type(sep).__name__ == 'str') and (not sep):
        return "seperator cannot be empty string"

    # handle no seperator
    if (type(sep).__name__ == 'bool'):
        result = []
        wordSep = False
        i = prev = 0
        while (not wordSep):
            if (str[i:i+1] != ' ') and (str[i:i+1] != '\n') and (str[i:i+1] != '\t') and (str[i:i+1] != '\r'):
                i += 1
            else:
                if prev != i:
                    result.append(str[prev:i])
                i += 1
                prev = i

            if i >= len(str):
                if prev != i:
                    result.append(str[prev:])
                wordSep = True

        return result

    # handle wrong value as seperator
    if (type(sep).__name__ != 'str'):
        return "seperator has to be a string or character"

    # handle limit = 0
    if (type(limit).__name__ == 'int') and limit == 0:
        return [str]

    result = []
    done = False
    i = prev = 0

    while (not done):
        if str[i:i+len(sep)] == sep:
            result.append(str[prev:i])
            i += len(sep)
            prev = i
        else:
            i += 1

        if (i >= len(str)) or (len(result) == limit):
            result.append(str[prev:])
            done = True

    return result

myStr = 'stringstreet'
myStr = 'stringstreetest'
myStr = 'string      street est'
myStr = 'string      street   '
myStr = '   my string      street   '
print(split(myStr, 'st'))
print(split(myStr, 's'))
print(split(myStr, 's', 0))
print(split(myStr, 's', 1))
print(split(myStr, 'xyz', 1))
print(split(myStr, 's', 2))
print(split(myStr, 's', 3))
print(split(myStr, 'st', 2))
print(split(myStr))
print(split(myStr, ' ', 3))

File: Chapter-04-Strings-AssociativeArrays/String-Split/String-Split.js
// Write your own function to recreate the string.split() function of javascript. split a string into an array of substrings. It takes in a seperator based on which to split the string.

function split(str, sep = -1, limit = -1){
  // if the seperator is not specified, return the entire string as an element in an array
  // the limit cannot be specified without specifying the seperator
  if (typeof(sep) === 'number'){
    return [str]
  }

  // the limit has to be a positive integer
  if ((typeof(limit) != 'number') || (limit == 0)){
    return []
  }
  // if seperator is empty string, break the entire string into individual characters and put them into the array
  var result = []
  var done = false
  var i = 0
  var prev = 0

  var compare = function(){
    if (str.substr(i, sep.length) == sep){
      result.push(str.substring(prev, i))
      i += sep.length
      prev = i
    } else {
      i++
    }

    if (i >= str.length){
      result.push(str.substring(prev, i))
      done = true
    }
  }

  if ((typeof(sep) === 'string') && (sep.length) && (limit > 0)){
    while ((result.length < limit) && (!done)){
      compare()
    }
  } else if ((typeof(sep) === 'string') && (sep.length)){
    while (!done){
      compare()
    }
  } else {
    for (pos in str){
      result.push(str.charAt(pos))
    }
  }

  return result
}

var myStr = "stringstreetest"
myStr = "stringstreet"
myStr = "string street"
console.log(`split('${myStr}',3)      -->`, split(myStr, 3))
console.log(`split('${myStr}')        -->`, split(myStr))
console.log(`split('${myStr}','st',5) -->`, split(myStr,'st',5))
console.log(`split('${myStr}','st',2) -->`, split(myStr,'st',2))
console.log(`split('${myStr}','st')   -->`, split(myStr,'st'))
console.log(`split('${myStr}','')     -->`, split(myStr,''))
console.log(`split('${myStr}','xyz')  -->`, split(myStr,'xyz'))
